src/mavlink-router/mainloop.cpp

```diff
@@ -310,16 +310,12 @@ int Mainloop::loop()
             if (events[i].events & EPOLLERR) {
                 log_error("poll error for fd %i", p->fd);
 
-                /*
-                 * TCP Pollables can be added/removed dynamically and
-                 * is_valid() is set when it becomes not available anymore.
-                 * Otherwise it's a unkonwn error or another bus has a
-                 * unexpected disconnect (e.g. when removing a usb-serial
-                 * device), so make it fatal: external components may restart
-                 * mavlink-router
-                 */
-                if (p->is_valid())
```

Comment:
the idea here was to reuse is_valid() because TCP endpoint is the only one to override this method and set it to  false if it got an error.

I would understand this solution if we were to move the EPOLLERR to be handled before EPOLLOUT/EPOLLIN. However since we are leaving it here, how is this different from what we currently have?

Reply:
Yes, I understood that `is_valid` was re-used. And, as you said, EPOLLOUT/EPOLLIN in handled before EPOLLERR, so I agree with your reasoning that when a socket is closed, the following would be expected:
- EPOLLIN should mark the fd as invalid
- EPOLLERR sees the invalid fd, and skips request exit

However, this is not what happens. During my testing, I discovered different behavior. Sometimes, EPOLLERR is handled without EPOLLIN. EPOLLIN happens the next iteration of the mainloop. So sometimes a closed TCP connection triggers EPOLLIN+EPOLLERR, sometimes only EPOLLERR. I do not know why this happens, but for this reason is_valid cannot be reused.